Adaptors/S3/src/ObjectStorage.cs

```diff
@@ -0,0 +1,249 @@
+  private readonly AmazonS3Client         s3Client_;
+
+  /// <summary>
+  ///   <see cref="IObjectStorage" /> implementation for S3
+  /// </summary>
+  /// <param name="s3Client">Connection to S3</param>
+  /// <param name="objectStorageName">Name of the object storage used to differentiate them</param>
+  /// <param name="options">S3 object storage options</param>
+  /// <param name="logger">Logger used to print logs</param>
+  public ObjectStorage(AmazonS3Client         s3Client,
+                       string                 objectStorageName,
+                       Options.S3             options,
+                       ILogger<ObjectStorage> logger)
+  {
+    s3Client_          = s3Client;
+    objectStorageName_ = objectStorageName;
+    options_           = options;
+    bucketName_        = options.BucketName;
+    logger_            = logger;
+  }
+
+  /// <inheritdoc />
+  public async Task AddOrUpdateAsync(string                                 key,
+                                     IAsyncEnumerable<ReadOnlyMemory<byte>> valueChunks,
+                                     CancellationToken                      cancellationToken = default)
+  {
+    using var _ = logger_.LogFunction(objectStorageName_ + key);
+
+    var idx      = 0;
+    var taskList = new List<Task>();
+    await foreach (var chunk in valueChunks.WithCancellation(cancellationToken)
+                                           .ConfigureAwait(false))
+    {
+      taskList.Add(s3Client_.WriteObjectAsync(bucketName_,
+                                              $"{objectStorageName_}{key}_{idx}",
+                                              chunk));
+      ++idx;
+    }
+
+    await s3Client_.WriteStringAsync(bucketName_,
+                                     $"{objectStorageName_}{key}_count",
+                                     idx.ToString())
+                   .ConfigureAwait(false);
+    await taskList.WhenAll()
+                  .ConfigureAwait(false);
+  }
+
+  /// <inheritdoc />
+  public async IAsyncEnumerable<byte[]> GetValuesAsync(string                                     key,
+                                                       [EnumeratorCancellation] CancellationToken cancellationToken = default)
+  {
+    using var          _        = logger_.LogFunction(objectStorageName_ + key);
+    GetObjectResponse? response = null;
+    try
+    {
+      response = await s3Client_.GetObjectAsync(bucketName_,
+                                                $"{objectStorageName_}{key}_count");
+    }
+    catch (AmazonS3Exception ex) when (ex.ErrorCode == "NoSuchKey")
+    {
+      logger_.LogError("The key {key} was not found.",
+                       key);
+      throw new ObjectDataNotFoundException("Key not found");
+    }
+
+    // Get the data from the response stream
+    using var reader      = new StreamReader(response.ResponseStream);
+    var       fileContent = reader.ReadToEnd();
+
+    var valuesCount = int.Parse(fileContent!);
```

Comment:
You should test if it is null and throw a proper exception if it is the case instead of the suppressing the nullable with `!`